### PR TITLE
Support quoted string and inline comment in config

### LINKED ISSUE #799: Unexpected behavior in configuration parsing
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Version

unstable

### Minimal reproduce step

In kvrocks.conf

```
somekey1 = value # comment
somekey2 = "value x" # comment
```

### What did you expect to see?

The parsed items:
```
{"somekey1": "value", "somekey1": "value x"}
```

### What did you see instead?

The parsed items:
```
{"somekey1": "value # comment", "somekey1": "\"value x\" # comment"}
```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by PragmaTwice**
Fix #799.

Refer to https://redis.io/docs/manual/config, now we have quoted strings as redis, and even more: we have inline comments.

Examples:
`key value1 value2 value3` -> `"key", "value1 value2 value3"`
`key " v\n "` -> `"key", " v\n "`
`key value1 # value2 # value3` -> `"key", "value1"`
`key "value1 # value2" # value3` -> `"key", "value1 # value2"`

### CODE REVIEW

**Review by caipengbo — DISMISSED** [review]

**Comment by PragmaTwice** [comment]
Hi everyone, I think this PR is ready to review now : )

File: src/config_util.cc [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+ * under the License.
+ *
+ */
+
+#include "config_util.h"
+#include "util.h"
+
+
+StatusOr<ConfigKV> ParseConfigLine(const std::string& line) {
+  enum { KEY, NORMAL, QUOTED, PRE_KEY_SPACE, AFTER_KEY_SPACE, AFTER_VAL_SPACE, ESCAPE, ERROR } state = PRE_KEY_SPACE;
```
Comment (tisonkun) [inline src/config_util.cc]: May we have a comment/document for this state machine?

**Review by tisonkun — COMMENTED** [review]
Generally looks good. One comment:

From https://redis.io/docs/manual/config/ it is valid:

```
keyword argument1 argument2 ... argumentN
```

but it seems in Kvrocks we support only one argument?

**Comment by PragmaTwice** [comment]
> Generally looks good. One comment:
> 
> From https://redis.io/docs/manual/config/ it is valid:
> 
> ```
> keyword argument1 argument2 ... argumentN
> ```
> 
> but it seems in Kvrocks we support only one argument?

I used to support multiple value configuration at the first (can see my first commit, it can parse multiple value configuration and even has a more simple state machine). But it involved too many changes, and the current configuration reading/dumping/rewriting/hot-loading mechanism in kvrocks do not handle multiple value configuration well, so I give up and move to support just one value configuration.

File: src/config_util.cc (hunk 1 above)
Comment (PragmaTwice) [reply src/config_util.cc]: done

File: src/config_util.cc [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+StatusOr<ConfigKV> ParseConfigLine(const std::string& line) {
+  enum {
+    KEY,  // in (unquoted) key string
+    NORMAL,  // in unquoted value string
+    QUOTED,  // in quoted value string
+    PRE_KEY_SPACE,  // in whitespace characters before key
+    AFTER_KEY_SPACE,  // in whitespace characters after key and before value
+    AFTER_VAL_SPACE,  // in whitespace characters after value
+    ESCAPE, // in escape character of quoted string
```
Comment (PragmaTwice) [inline src/config_util.cc]:
```suggestion
    ESCAPE,  // in escape character of quoted string
```

**Comment by tanruixiang** [comment]
Hi. When we parse `"a''b"`, May I ask which of the following is in line with expectations.
`"a''b"` => `"ab"`
`"a''b"` => `"a''b"`
The code parses out the second case.

**Comment by PragmaTwice** [comment]
> Hi. When we parse `"a''b"`, May I ask which of the following is in line with expectations. `"a''b"` => `"ab"` `"a''b"` => `"a''b"` The code parses out the second case.

if you mean `key "a''b"`, the result will be `key`, `a''b`.

> `"a''b"` => `"ab"`

IMHO, I have never seen this behavior, it is weird.

**Comment by tanruixiang** [comment]
> > Hi. When we parse `"a''b"`, May I ask which of the following is in line with expectations. `"a''b"` => `"ab"` `"a''b"` => `"a''b"` The code parses out the second case.
> 
> if you mean `key "a''b"`, the result will be `key`, `a''b`.

Yes. I mean Is it necessary to treat `''` as null？For the current code, parse `"a\'\'b"`'s reusult is equal to  `"a''b"`.

**Comment by PragmaTwice** [comment]
> > > Hi. When we parse `"a''b"`, May I ask which of the following is in line with expectations. `"a''b"` => `"ab"` `"a''b"` => `"a''b"` The code parses out the second case.
> > 
> > 
> > if you mean `key "a''b"`, the result will be `key`, `a''b`.
> 
> Yes. I mean Is it necessary to treat `''` as null？For the current code, parse `"a\'\'b"`'s reusult is equal to `"a''b"`.

There is no programming language / shell language / markup language treat `'` as null in a quoted string started with `"`.

**Comment by tanruixiang** [comment]
> 



> > > > Hi. When we parse `"a''b"`, May I ask which of the following is in line with expectations. `"a''b"` => `"ab"` `"a''b"` => `"a''b"` The code parses out the second case.
> > > 
> > > 
> > > if you mean `key "a''b"`, the result will be `key`, `a''b`.
> > 
> > 
> > Yes. I mean Is it necessary to treat `''` as null？For the current code, parse `"a\'\'b"`'s reusult is equal to `"a''b"`.
> 
> There is no programming language / shell language / markup language treat `'` as null in a quoted string started with `"`.

Sorry, I made a mistake. Hope it didn't bother you

**Comment by PragmaTwice** [comment]
> Sorry, I made a mistake. Hope it didn't bother you

never mind, feel free to find problems and ask questions.

**Review by tisonkun — APPROVED** [review]

**Comment by PragmaTwice** [comment]
Hi all, I will merge it if no further discussion, feel free to review it.

**Comment by git-hulk** [comment]
> Hi all, I will merge it if no further discussion, feel free to review it.

Let's me have a test locally, will feedback soon

**Review by git-hulk — APPROVED** [review]

**Pull request merged**